### ISSUE TITLE
Query: Revisit caching synchronization mechanism

### ISSUE THREAD

**Issue by smitpatel** [dotnet/efcore#18516]
In QueryCompiler
In RelationalCommandCache
For full discussion see https://github.com/aspnet/EntityFrameworkCore/pull/18497#discussion_r337291324

**Comment by ajcvickers** [comment]
Note from triage: the main concern here is that if we allow a race when, for example, compiling the model, then that compilation needs to be thread safe. If it isn't, then we must block two threads from attempting to compile at the same time.

**Comment by roji** [comment]
Here is some analysis/thoughts about QueryCompiler. It's my first time in this area so I'm probably missing context etc.

tl;dr I don't think we have a bug, but we do have a non-ideal spinning loop which should be fixed. Also not sure the synchronization in question is necessary.

* Some services participating in compilation are scoped (e.g. IQueryCompiler, IDatabase), while others are instantiated per-compilation via singleton factories (e.g. QueryableMethodTranslatingExpressionVisitor). If I'm reading this right, that means there's currently no need for us to do any synchronization/blocking/locking, since DbContext itself isn't thread-safe.
* In CompiledQueryCache we have a synchronization mechanism based [on the query cache key](https://github.com/aspnet/EntityFrameworkCore/blob/fce44c324f6c096cce5457b9bd0b75a2b1ade3da/src/EFCore/Query/Internal/CompiledQueryCache.cs#L71). I don't think is a safety mechanism - it would be strange to allow two compilations to proceed at the same time, except when they happen to have the same key (and once again, such protection doesn't seem necessary as compilation services are scoped).
* This synchronization mechanism also has a race condition: one thread may have gotten false from `_memoryCache.TryGetValue` and entered the condition, just as another thread completed compilation and removed the sync object. In this case the 1st thread will proceed to recompile.
* So I'm guessing this was meant as some sort of optimization, where if a query is already being compiled, other threads will block until it is done rather than doing the same compilation again.
* I'm not sure this optimization is worth it - it seems to work on a specific timing case, which intuitively I'm not sure is wide-spread: do we think applications typically perform the same query from multiple threads at the same time at startup?
* Even if we want to keep this optimization, the synchronization mechanism itself should not involve a spinning loop: threads waiting use 100% CPU. This could be rewritten with a locking mechanism instead (although again I'm not sure the optimization makes sense in general).

I think a lot of the above applies also to RelationalCommandCache, but I haven't looked at it.

**Comment by roji** [comment]
As discussed with @AndriySvyryd and @smitpatel I'll do a bit of benchmarking around this to measure the impact of this optimization. However, I'm skeptical that we'd get much useful data from here: it all depends on our assumptions on multiple threads compiling the same query around the same time at app startup.
